tests/buffer_test.py

```diff
@@ -81,6 +75,109 @@ def test_config_buffer_profile_add_headroom_to_dynamic_profile(self):
+
+    def test_config_shp_ratio_negative(self):
+        runner = CliRunner()
+        db = Db()
+        port_number = len(db.cfgdb.get_table('PORT').keys())
+        bad_oversubscribe_ratio = str(port_number + 1)
+        result = runner.invoke(config.config.commands["buffer"].commands["shared-headroom-pool"].commands["over-subscribe-ratio"],
+                               [bad_oversubscribe_ratio], obj=db)
+        print(result.exit_code)
+        print(result.output)
+        assert result.exit_code != 0
+        assert "Invalid over-subscribe-ratio value {}. It should be in range [0, {}]".format(bad_oversubscribe_ratio, port_number) in result.output
+
+    def test_config_buffer_profile_headroom_toggle_shp(self):
+        runner = CliRunner()
+        db = Db()
+
+        # disable SHP by setting over-subscribe-ratio to 0
+        result = runner.invoke(config.config.commands["buffer"].commands["shared-headroom-pool"].commands["over-subscribe-ratio"],
+                               ["0"], obj=db)
+        print(result.exit_code)
+        print(result.output)
+        assert result.exit_code == 0
+        profile = db.cfgdb.get_entry('DEFAULT_LOSSLESS_BUFFER_PARAMETER', 'AZURE') == {'default_dynamic_th': '0', 'over_subscribe_ratio': '0'}
+
+        # size should equal xon + xoff
+        result = runner.invoke(config.config.commands["buffer"].commands["profile"].commands["add"],
+                               ["test1", "--dynamic_th", "3", "--xon", "18432", "--xoff", "32768"], obj=db)
+        print(result.exit_code)
+        print(result.output)
+        assert result.exit_code == 0
+        profile = db.cfgdb.get_entry('BUFFER_PROFILE', 'test1')
+        assert profile == {'dynamic_th': '3', 'pool': '[BUFFER_POOL|ingress_lossless_pool]', 'xon': '18432', 'xoff': '32768', 'size': '51200'}
+
+        # xoff should equal size - xon
+        result = runner.invoke(config.config.commands["buffer"].commands["profile"].commands["add"],
+                               ["test2", "--dynamic_th", "3", "--xon", "18432", "--size", "32768"], obj=db)
+        print(result.exit_code)
+        print(result.output)
+        assert result.exit_code == 0
+        profile = db.cfgdb.get_entry('BUFFER_PROFILE', 'test2')
+        assert profile == {'dynamic_th': '3', 'pool': '[BUFFER_POOL|ingress_lossless_pool]', 'xon': '18432', 'xoff': '14336', 'size': '32768'}
+
+        # set size
+        result = runner.invoke(config.config.commands["buffer"].commands["profile"].commands["set"],
+                               ["test2", "--dynamic_th", "3", "--size", "65536"], obj=db)
+        print(result.exit_code)
+        print(result.output)
+        assert result.exit_code == 0
+        profile = db.cfgdb.get_entry('BUFFER_PROFILE', 'test2')
+        assert profile == {'dynamic_th': '3', 'pool': '[BUFFER_POOL|ingress_lossless_pool]', 'xon': '18432', 'xoff': '14336', 'size': '65536'}
```

Comment:
To **set** a profile means to modify an existing profile. In this case, only the parameters explicitly provided will be updated.
If a new profile is being added, we have this assumption in case `size` is not provided and SHP is disabled.